staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -0,0 +1,427 @@
+var buildDefinitions sync.Once
+var namer *openapi.DefinitionNamer
+
+// BuildSwagger builds swagger and calculates etag for given crd in given version
+func BuildSwagger(crd *apiextensions.CustomResourceDefinition, version string) (*spec.Swagger, string, error) {
+	start := time.Now()
+	var schema *spec.Schema
+	s, err := apiextensions.GetSchemaForVersion(crd, version)
+	if err != nil {
+		return nil, "", err
+	}
+	if s != nil && s.OpenAPIV3Schema != nil {
+		schema, err = ConvertJSONSchemaPropsToOpenAPIv2Schema(s.OpenAPIV3Schema)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+	b := newBuilder(crd, version, schema)
+
+	// Sample response types for building web service
+	sample := &CRDCanonicalTypeNamer{
+		group:   b.group,
+		version: b.version,
+		kind:    b.kind,
+	}
+	sampleList := &CRDCanonicalTypeNamer{
+		group:   b.group,
+		version: b.version,
+		kind:    b.listKind,
+	}
+	status := &metav1.Status{}
+	patch := &metav1.Patch{}
+	scale := &v1.Scale{}
+
+	routes := make([]*restful.RouteBuilder, 0)
+	root := fmt.Sprintf("/apis/%s/%s/%s", b.group, b.version, b.plural)
+	if b.namespaced {
+		routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList).
+			Operation("list"+b.kind+"ForAllNamespaces"))
+		root = fmt.Sprintf("/apis/%s/%s/namespaces/{namespaces}/%s", b.group, b.version, b.plural)
+	}
+	routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList))
+	routes = append(routes, b.buildRoute(root, "", "POST", "create", sample).Reads(sample))
+	routes = append(routes, b.buildRoute(root, "", "DELETE", "deletecollection", status))
+
+	routes = append(routes, b.buildRoute(root, "/{name}", "GET", "read", sample))
+	routes = append(routes, b.buildRoute(root, "/{name}", "PUT", "replace", sample).Reads(sample))
+	routes = append(routes, b.buildRoute(root, "/{name}", "DELETE", "delete", status))
+	routes = append(routes, b.buildRoute(root, "/{name}", "PATCH", "patch", sample).Reads(patch))
+
+	subresources, err := apiextensions.GetSubresourcesForVersion(crd, version)
+	if err != nil {
+		return nil, "", err
+	}
+	if subresources != nil && subresources.Status != nil {
+		routes = append(routes, b.buildRoute(root, "/{name}/status", "GET", "read", sample))
+		routes = append(routes, b.buildRoute(root, "/{name}/status", "PUT", "replace", sample).Reads(sample))
+		routes = append(routes, b.buildRoute(root, "/{name}/status", "PATCH", "patch", sample).Reads(patch))
+	}
+	if subresources != nil && subresources.Scale != nil {
+		routes = append(routes, b.buildRoute(root, "/{name}/scale", "GET", "read", scale))
+		routes = append(routes, b.buildRoute(root, "/{name}/scale", "PUT", "replace", scale).Reads(scale))
+		routes = append(routes, b.buildRoute(root, "/{name}/scale", "PATCH", "patch", scale).Reads(patch))
+	}
+
+	for _, route := range routes {
+		b.ws.Route(route)
+	}
+
+	elapsed := time.Since(start)
+	klog.Errorf(">>>>> prepare crd openapi before build took %s", elapsed)
```

Comment:
Remove this before merge. if this is concern, we should add a monitoring metric for it.

Reply:
ack. it's been addressed in https://github.com/kubernetes/kubernetes/pull/71223